src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,237 @@
+                    // is there any effect on the view?
+                    if (viewLines.IntersectsBufferSpan(changedSpan))
+                    {
+                        AdornmentLayer.RemoveAdornmentsByVisualSpan(changedSpan);
+                    }
+                }
+            }
+
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            var tagSpanToPointMap = GetTagSpansToSnapshotPointMap(map);
+            foreach (var tagMappingSpanList in map.Values)
+            {
+                var tagMappingSpan = GetHighestOrderTag(tagMappingSpanList);
+                TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span);
+                var geometry = viewLines.GetMarkerGeometry(span);
+                if (geometry != null)
+                {
+                    var tag = tagMappingSpan.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.TagID + tag.ErrorType);
+                    var graphicsResult = tag.GetGraphics(TextView, geometry, GetFormat(classificationType));
+                    if (!tagSpanToPointMap.TryGetValue(tagMappingSpan, out var point))
+                    {
+                        continue;
+                    }
+
+                    var lineView = TextView.GetTextViewLineContainingBufferPosition(point);
+
+                    var visualElement = graphicsResult.VisualElement;
+                    if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfCode)
+                    {
+                        Canvas.SetLeft(visualElement, lineView.Right);
+                    }
+                    else if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+                    {
+                        Canvas.SetLeft(visualElement, TextView.ViewportWidth - visualElement.DesiredSize.Width);
+                    }
+
+                    tag.EditorTextLocation = lineView.Right;
+                    Canvas.SetTop(visualElement, geometry.Bounds.Bottom - visualElement.DesiredSize.Height);
+
+                    if (lineView.Right < TextView.ViewportWidth - visualElement.DesiredSize.Width)
+                    {
+                        AdornmentLayer.AddAdornment(
+                            // TextRelative because the adornment is positioned relative to the text in the editor
+                            behavior: AdornmentPositioningBehavior.TextRelative,
+                            visualSpan: span,
+                            tag: tag,
+                            adornment: visualElement,
+                            removedCallback: delegate { graphicsResult.Dispose(); });
+                    }
+                }
+            }
+        }
+
+        private static Dictionary<IMappingTagSpan<InlineDiagnosticsTag>, SnapshotPoint> GetTagSpansToSnapshotPointMap(
+            IDictionary<(int linenum, SnapshotSpan changedSpan), List<IMappingTagSpan<InlineDiagnosticsTag>>> map)
```

Comment:
what is this map used for?  i can't figure out what the point of it is.

Reply:
The point was to map a span back to a point to be used for getting the IWPFTextViewLine, but it was removed.